force-app/main/default/classes/DMLOperations.cls

```diff
@@ -146,8 +160,65 @@ public with sharing class DMLOperations {
+
+        // Case 1: Found existing Account with the given accountName -> assign Account Id to accountRecord variable
+        // Case 2: Create a new Account record with the given accountName
+        List<Account> accountList = [SELECT Id FROM Account WHERE Name = :accountName Limit 1];
+        Account accountRecord = new Account();
+
+        if (!accountList.isEmpty()) {
+            accountRecord.Id = accountList.get(0).Id;
+        } else {
+            accountRecord.Name = accountName;
+        }
+        upsert accountRecord;
+
+        // Case 1a: Some Opportunity records (existingOpportunities) already exist with the given names -> Assign accountRecord.Id to those opportunities
+        // Case 1b: Some Opportunity records (newOppRecords) do not exist with the given names -> Create new Opportunity records with the given names -> Assign accountRecord.Id to new opportunities 
+        // Case 2: For all the given OppNames, create new opportunities -> Assign accountRecord.Id to those opportunities
+        List<Opportunity> existingOpportunities = new List<Opportunity>();
+        if (!oppNames.isEmpty()) {
+            existingOpportunities = [SELECT Id, Name, AccountId FROM Opportunity WHERE Name IN :oppNames];
+        }
+        List<String> existingOppNames = new List<String>();
+        List<String> newOppNames = new List<String>();
+        List<Opportunity> newOppRecords = new List<Opportunity>();
+        if (!existingOpportunities.isEmpty()) {
+            for (Opportunity currentOppRecord : existingOpportunities) {
+                if (currentOppRecord.AccountId != accountRecord.Id) {
+                    currentOppRecord.AccountId = accountRecord.Id;
+                }
+                existingOppNames.add(currentOppRecord.Name);
+            }     
+            upsert existingOpportunities;
+            for (String currentOppName1 : oppNames) {
+                if (!existingOppNames.contains(currentOppName1)) {
+                    newOppNames.add(currentOppName1);
+
+                }
+            }
+            for (String newOppName : newOppNames) {
+                Opportunity newOppRecord = new Opportunity();
+                newOppRecord.AccountId = accountRecord.Id;
+                newOppRecord.Name = newOppName;
+                newOppRecord.StageName = 'Prospecting';
+                newOppRecord.CloseDate = System.today()+30;
+                newOppRecords.add(newOppRecord);
+            }
+            upsert newOppRecords;      
+        }else {
+            for (String currentOppName2 : oppNames) {
+                Opportunity newOppRecord = new Opportunity();
+                newOppRecord.AccountId = accountRecord.Id;
+                newOppRecord.Name = currentOppName2;
+                newOppRecord.StageName = 'Prospecting';
+                newOppRecord.CloseDate = System.today()+30;
+                newOppRecords.add(newOppRecord);
+            }
+            upsert newOppRecords;
```

Comment:
My advice for this would to look for ways to combine the construction of the new Opportunity into one block. Right now you are duplicating setting the same values into an Opportunity. If you need to change that logic, you now have to change in 2 places. Always try to not repeat yourself.